### PR TITLE
⚠️ ImageFilter - add exclusive validation -> pointers

### PR DESCRIPTION
**Pull request by dulek**
**What this PR does / why we need it**:
This commit changes `ID` and `Name` of `ImageFilter` to pointers which should only affect go marshalling.

Other than that it adds CEL validation of the ImageFilter, so that Name or Tags can only be set when ID is unset. Conversions are updated accordingly to make sure we only set Name when ID is unset.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [X] squashed commits
- if necessary:
  - [X] includes documentation
  - [X] adds unit tests

/hold

### CODE REVIEW

File: api/v1beta1/types.go [hunk 1]
```diff
@@ -28,12 +28,19 @@ type OpenStackMachineTemplateResource struct {
 	Spec OpenStackMachineSpec `json:"spec"`
 }
 
+// ImageFilter describes the data needed to identify which image to use.
+// +kubebuilder:validation:XValidation:rule="(has(self.id) && !has(self.name) && !has(self.tags)) || !has(self.id)",message="when ID is set you cannot set other options"
```
Comment (mdbooth) [inline api/v1beta1/types.go:32]: Shiny!

File: api/v1beta1/types.go [hunk 2]
```diff
@@ -28,12 +28,19 @@ type OpenStackMachineTemplateResource struct {
...
 type ImageFilter struct {
-	// The ID of the desired image. If this is provided, the other filters will be ignored.
-	ID string `json:"id,omitempty"`
+	// The ID of the desired image. If ID is provided, the other filters cannot be provided.
+	// +kubebuilder:validation:Format:=uuid
```
Comment (mdbooth) [inline api/v1beta1/types.go:35]: Please can you also add a test for uuid format validation?

File: test/e2e/suites/apivalidations/filters_test.go
Comment (mdbooth) [inline test/e2e/suites/apivalidations/filters_test.go:1]: Nice! Please can you add one more test that an invalid uuid is rejected?

File: api/v1beta1/openstackmachine_types.go [hunk 3]
```diff
@@ -43,7 +43,7 @@ type OpenStackMachineSpec struct {
 
 	// The image to use for your server instance.
 	// If the rootVolume is specified, this will be used when creating the root volume.
-	Image ImageFilter `json:"image,omitempty"`
+	Image ImageFilter `json:"image"`
```
Comment (mdbooth) [inline api/v1beta1/openstackmachine_types.go:46]: Please can you add an explicit `// +optional`?

**Review by mdbooth — APPROVED** [review]
Nice! Approved, but I'd appreciate it if you could add the additional requested test and fix the nit.

File: api/v1beta1/types.go [hunk 4]
```diff
@@ -28,12 +28,19 @@ type OpenStackMachineTemplateResource struct {
 	Spec OpenStackMachineSpec `json:"spec"`
 }
 
+// ImageFilter describes the data needed to identify which image to use.
```
Comment (mdbooth) [inline api/v1beta1/types.go:31]: One more thing: please can you summarise the validation requirements below in this user-visible godoc?

File: api/v1beta1/openstackmachine_types.go (hunk 3 above)
Comment (dulek) [reply api/v1beta1/openstackmachine_types.go:46]: I wasn't exactly sure if the image is optional. Is it? Should I make it `Image *ImageFilter` then?

File: test/e2e/suites/apivalidations/filters_test.go
Comment (dulek) [reply test/e2e/suites/apivalidations/filters_test.go:1]: There's one here: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1939/files#diff-e5adab9fc4945c556b3da78bd7188014fc1150d23e2be34d46e09e088eeb131aR206-R212

**Comment by dulek** [comment]
@mdbooth: I fixed a comment, but I'm still unsure about `ImageFilter` being optional. Digging into that now.

File: api/v1beta1/openstackmachine_types.go (hunk 3 above)
Comment (dulek) [reply api/v1beta1/openstackmachine_types.go:46]: I think we can't make `ImageFilter` optional, code requires that the exactly one image will be found. An empty `ImageFilter` might be possible and resolvable if you have just one image in that tenant.

Comment (mdbooth) [reply api/v1beta1/openstackmachine_types.go:46]: D'oh. Of course this should not be optional. Yep, lets make it explicitly required.

**Comment by EmilienM** [comment]
/test pull-cluster-api-provider-openstack-e2e-full-test
/lgtm

I think all comments were addressed (nice work!) but I'll let Matt hold cancel.

**Comment by dulek** [comment]
I added the explicit `+required` on the `ImageFilter`. I guess this is good to go, I can expand the filter in the Bastion patch.

**Pull request merged**